[TF:TRT] Reorganize testing utilities.

christopherbate opened the pull request:
Creates TF-TRT "testutils" library and replaces some utility functions with gMock matchers. Renames GetTestDims to CreateDims. Adds ostream<< printers for nvinfer1 types. Replaces "const char*" usage in test fixtures with "std::string" types.
Moves two std::vector templated utility functions `InitTestVector` and `CastTestVector` to `trt_testutils.h` from `convert_nodes_test.cc`, and renamed to `CreateVectorIota` and `CastVector` respectively.

christopherbate committed:
TF-TRT - Refactor test utilities

christopherbate commented:
@tfeher

christopherbate commented:
@bixia1 Can you help review? Thanks

tfeher commented:
Thanks for this PR @christopherbate, it is nice to see some clean up in the test routines! In an offline discussion you have mentioned that "replacing ExpectsX functions with GMock matchers eliminates indirection in the test error result so the test error reports the actual failing line in the test [...] and correctly prints the nvinfer types." Could you update the PR description to include this information?

christopherbate commented on tensorflow/compiler/tf2tensorrt/utils/trt_testutils.h:
```diff
@@ -0,0 +1,178 @@
+#include "tensorflow/core/lib/core/status_test_util.h"
+#include "third_party/tensorrt/NvInfer.h"
+
+namespace nvinfer1 {
```
I think I should move this to common/utils.h as it is useful outside of testing.
Also I will add == operator for `nvinfer1::Dims` using the same traits filter.

christopherbate commented on tensorflow/compiler/tf2tensorrt/utils/trt_testutils.h:
```diff
@@ -0,0 +1,178 @@
+using enable_if_nvinfer_dims =
+    std::enable_if<std::is_base_of<nvinfer1::Dims, T>::value, T>;
+
+// Prints nvinfer1::Dims (and any sub-struct, for legacy Dim types)
+// to ostream.
```
I say legacy because those derived types are now deprecated since an older version of TRT. I think it's TRT 5, but would have to double check.

christopherbate commented:
@bixia1,
Updated based on your comments, and fixed a subtle issue with the TRT_ShapedWeights gMock matcher. Tests look good to go on my end.
I have a follow on PR for additional test refactoring which I'm about to create. I can either create a new PR or append to this one? Thanks.

bixia1 commented on tensorflow/compiler/tf2tensorrt/common/utils.h:
```diff
@@ -54,6 +56,64 @@ void MaybeInitializeTrtPlugins(nvinfer1::ILogger* trt_logger);
 
+namespace nvinfer1 {
+
+// Alias a useful type computation for nvinfer1::Dims types. Types like
+// nvinfer1::Dims2 inherit from nvinfer1::Dims.
```
I think we should remove the support for the legacy DimsN for two reasons: We don't actually use it in the TF-TRT converter. I see a few use in the test, but that can be fixed; DimsN is a convenient way of defining Dims (fix the nbDims to N) and will be deprecated in TensorRT. If we only support Dims NOT support DimsN, will this simplify the code here?

christopherbate replied:
In a later PR I can remove deprecated type usages from the tests, and then remove this SFINAE template usage. 

If I understand what you're saying, we should have utilities called DimsN, which would just be functions which create an `nvinfer1::Dims`?

In follow on [PR I moved utility functions](https://github.com/tensorflow/tensorflow/pull/51435/files#diff-bd883d225ea57894fd8bbe9160ac8cc3233c1947bb532fb544c309d7dcaecb03R138-R154) under `common/utils.h` for nvinfer1::Dims creation; I can put the DimsN creation utilities there later when I remove the deprecated types.

bixia1 replied:
Can you explain why you need to use SFINAE here?
What would be the problem if you just define std::ostream& operator<<(std::ostream& os, const nvinfer1::Dims& v) in this PR instead of introducing SFINAE and then remove it later? 
We don't actually use DimsN in the converter. But even if we use DimsN, std::ostream& operator<<(std::ostream& os, const nvinfer1::Dims& v) should also works for DimsN, am I right here?

christopherbate replied:
Yes it works that way as well. I just like this template technique because you can apply it to to things like non-ref values and the value type of a vector, etc. I feel like it tends to come up a lot in header files like this for stream operators and looks cleaner when they all have template definition. Up to you, I'm fine either way

bixia1 replied:
We only use SFINAE when it is necessary. Please simplify this code and remove SFINAE then.

bixia1 commented on tensorflow/compiler/tf2tensorrt/utils/trt_testutils.cc:
```diff
@@ -0,0 +1,84 @@
+  return node_def;
+}
+
+bool TrtDimsEquals(const nvinfer1::Dims& lhs, const nvinfer1::Dims& rhs) {
+  if (lhs.nbDims != rhs.nbDims) return false;
+  for (int i = 0; i < lhs.nbDims; ++i) {
+    if (lhs.d[i] != rhs.d[i]) return false;
+  }
+  return true;
+}
```
Can we replace this with the == operator added in this PR?

bixia1 commented on tensorflow/compiler/tf2tensorrt/utils/trt_testutils.h:
```diff
@@ -0,0 +1,193 @@
+#include "tensorflow/cc/framework/ops.h"
+#include "tensorflow/cc/framework/scope.h"
+#include "tensorflow/cc/ops/standard_ops.h"
+#include "tensorflow/compiler/tf2tensorrt/common/datavec.h"
```
I don't see any we need this header here. Can you try to eliminate the unnecessary hearders in this file?

christopherbate commented:
@bixia1 I applied your asks/recommendations.

bixia1 commented on tensorflow/compiler/tf2tensorrt/utils/trt_testutils.h:
```diff
@@ -0,0 +1,178 @@
+
+// nvinfer1::Dims GMock matchers
+
+// matches nvinfer1::Dims to initializer list or vector of ints
```
NIT: s/matches/Matches/
Would you please also add a period at the end of this line?

bixia1 reviewed (commented):
Would you please also do these before I will approve the PR:
(1) squash the commits into one commit. This is because we don't want the commit history for this PR to show in the final change, to simplify the commit message for the change.
(2) Would you please revise/simplify your PR description, such as removing "Added (8/10)" etc? This is because the PR description will become part of commit message, and we only want to keep the description about the final code change in the PR.

christopherbate commented:
> Would you please also do these before I will approve the PR:
> (1) squash the commits into one commit. This is because we don't want the commit history for this PR to show in the final change, to simplify the commit message for the change.
> (2) Would you please revise/simplify your PR description, such as removing "Added (8/10)" etc? This is because the PR description will become part of commit message, and we only want to keep the description about the final code change in the PR.

Will do, thanks

gbaned commented:
@bixia1 Can you please take a look on above comments from @christopherbate. Thanks!

bixia1 commented:
@gbaned I am still waiting for @christopherbate to address https://github.com/tensorflow/tensorflow/pull/51235#issuecomment-899784046

christopherbate commented:
@bixia1 done!

gbaned commented:
@christopherbate Can you please address Ubuntu Sanity errors? Thanks!

christopherbate commented:
@gbaned I applied the builifier recommended changes.

christopherbate commented:
Rebased on master

christopherbate commented:
@bixia1 @gbaned Please merge these queued dynamic shape support PRs first, then I will rebase this if required:

https://github.com/tensorflow/tensorflow/pull/51468
https://github.com/tensorflow/tensorflow/pull/51462

bixia1 commented on tensorflow/compiler/tf2tensorrt/BUILD:
```diff
@@ -99,6 +99,43 @@ cc_library(
+        ":testutils",
+        "//tensorflow/core:test_main",
+        "//tensorflow/core:protos_all_cc",
+        "@com_google_protobuf//:protobuf",
```
This line doesn't seem to work. Can you use this
https://github.com/tensorflow/tensorflow/blob/35571d62233eacb329f794aaa4ebe80d8720c3bf/tensorflow/core/BUILD#L321

"//tensorflow/core/platform:protobuf",

bixia1 commented on tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc:
```diff
@@ -440,7 +262,7 @@ TEST(TRT_TensorOrWeights_Test, Basic) {
       EXPECT_EQ(true, ptr->is_weights());
       EXPECT_TRUE(TrtShapedWeightsEquals(weights, ptr->weights()));
-      ExpectTrtDimsEqualsArray({}, ptr->GetTrtDims());
+      EXPECT_THAT(ptr->GetTrtDims(), DimsAreArray({}));
```
third_party/tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc:264:51: error: call to implicitly-deleted default constructor of 'tensorflow::tensorrt::convert::DimsAreArrayHelperMatcherP<std::vector<int>> &&'
      EXPECT_THAT(ptr->GetTrtDims(), DimsAreArray({}));

There are similar cases in other places.

bixia1 replied:
I am not sure why you can't reproduce this, but replacing the empty vector "{}" by "empty_vector" defined as follow fix the problem.
std::vector<int> empty_vector;

christopherbate replied:
Sure I can do that. It's pretty clear from all the CI builds that the public Tensorflow build does not have this problem. The only builds failing are Windows, which fails on some unrelated issue, and the "feedback/copybara", which I cannot view.

christopherbate replied:
Do you have "-fno-elide-constructors" set?

christopherbate commented:
@bixia1 I added the suggested change for the empty dims `DimsAreArray({})`

bixia1 commented on tensorflow/compiler/tf2tensorrt/utils/trt_testutils_test.cc:
```diff
@@ -0,0 +1,98 @@
+  EXPECT_THAT(nvinfer1::Dims4(1, 2, 3, 4), DimsAreArray({1, 2, 3, 4}));
+  // Check empty dims.
+  EXPECT_THAT(nvinfer1::Dims{}, Not(DimsAreArray({1, 2})));
+  EXPECT_THAT(nvinfer1::Dims{}, DimsAreArray({}));
```
Need to replace DimsAreArray({}) with a local as you did to another test.

bixia1 replied:
@christopherbate
Would you please fix this?

bixia1 replied:
I will fixed this when I merged it.
